samples/msal-node-samples/on-behalf-of-distributed-cache/src/CustomCachePlugin.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ICachePlugin, TokenCacheContext } from "@azure/msal-common";
+import { ICacheClient } from "@azure/msal-node";
+import { performance } from "perf_hooks";
+
+/**
+ * This class implements the ICachePlugin interface, which allows for the
+ * customization of the token cache. For more information, visit:
+ * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/caching.md
+ */
+class CustomCachePlugin implements ICachePlugin {
+    private client: ICacheClient;
+    private partitionKey: string;
+
+    constructor(client: ICacheClient, partitionKey: string) {
+        this.client = client;
+        this.partitionKey = partitionKey;
+    }
+
+    public async beforeCacheAccess(cacheContext: TokenCacheContext): Promise<void> {
+        performance.mark("beforeCacheAccess-start");
+        const cacheData = await this.client.get(this.partitionKey);
+        cacheContext.tokenCache.deserialize(cacheData);
+        performance.mark("beforeCacheAccess-end");
+        performance.measure("beforeCacheAccess", "beforeCacheAccess-start", "beforeCacheAccess-end");
+    }
+
+    public async afterCacheAccess(cacheContext: TokenCacheContext): Promise<void> {
+        if (cacheContext.cacheHasChanged) {
```

Comment:
Is there any information about the tokens within? You can set the TTL in Redis to `access_token.expiration` here.

Reply:
looking into it

Reply:
@bgavrilMS coming back to this. We are caching the entire cache blob for this user, so the cache contains id, access and refresh tokens. And potentially there could be a number of access tokens for different targets/resources. The TTL here however is for the entire cache blob -so I'm not sure how we could be accurate given that we have different auth artifacts. Any thoughts?

Reply:
@bgavrilMS friendly ping

Reply:
In MSAL.NET we suggest the TTL as being:

- max (expiry (AT)) if only access tokens are in teh cache
- no suggestion if at least one RT is in the cache

Reply:
But anyway, I think this is difficult without MSAL help.

Reply:
Ah, ok. All of our user flows (if not mistaken) includes a refresh token (due to default offline_access scope). Leaving this out for now then -we can revisit later.

Reply:
@derisen - client_credentials does not use RT

Reply:
Gotcha! -I made a note for client credentials, we can update the sample with the max (expiry (AT)) logic for TTL that you suggested.